### DISCUSSION TITLE
Drag-n-Drop files to/from FAR

### DISCUSSION THREAD

**Discussion by faerot** [elfmz/far2l#2399] [General]
My solution:
Build "dragon" program from here https://github.com/mwh/dragon and put it somewhere, i.e. to ``~/.config/far2l/dragon``
Also wrote a script and put there ``~/.config/far2l/dragon.py``:
```python
#!/usr/bin/python3

import os
import shutil
import subprocess 
import sys
from urllib.parse import unquote
from urllib.request import urlretrieve

if len(sys.argv) <= 1:
    print('\nUSAGE: dragon.py /destination/dir/\n')
else:
    dst_dir = os.path.expanduser(sys.argv[1])
    if not os.path.isdir(dst_dir):
        print('ERROR: ', dst_dir, ' is not a directory')
        sys.exit(1)

    this_dir = os.path.dirname(os.path.abspath(__file__))
    p = subprocess.Popen([os.path.join(this_dir, 'dragon'),
                          '--target', '--on-top', '--and-exit'], 
                         stdin=subprocess.PIPE, stdout=subprocess.PIPE)
    stdout, stderr = p.communicate()
    for url in stdout.decode().splitlines():
        dst_fn = os.path.join(dst_dir, os.path.basename(unquote(url)))
        if os.path.exists(dst_fn):
            print('ERROR: {} - already exists, skipped!'.format(dst_fn))
            continue
        print('+ ' + dst_fn + ' <== ' + url)
        if url.startswith('file://'):
            shutil.copy2(unquote(url[len('file://'):]), dst_fn)
        else:
            try:
                urlretrieve(url, dst_fn)
            except Exception as e:
                print('  ^^^ ' + str(e))
```

Add following menu items to user menu:
To drop:
![image](https://github.com/user-attachments/assets/381a98f2-fd2b-411b-ba73-8441b9ae5a88)

To drag:
![image](https://github.com/user-attachments/assets/ec496378-c6d9-440a-b9ff-d87832c6ffc7)

Drawback: extra steps - invoke menu items first, but better than nothing.
How do you solve this? Do you have better solutions?

**Comment by unxed** [comment]
Issue exists long time ago. Nothing was actually done AFAIK.
https://github.com/elfmz/far2l/issues/555

**Comment by m32** [comment]
#2602 - copy/paste files with gnome support

**Reply by faerot** [reply]
Great, what about KDE?

**Reply by m32** [reply]
it all depends on how the clipboard works in KDE, I don't have it and it's hard for me to check
